src/Reflex/Process.hs

```diff
@@ -127,27 +146,38 @@ createProcess
   => CreateProcess
   -> ProcessConfig t ByteString
   -> m (Process t ByteString ByteString)
-createProcess = createRedirectedProcess input output output
-  where
+createProcess p processConfig = do
+  channel <- liftIO newTChanIO
+
+  let
     input h = do
       H.hSetBuffering h H.NoBuffering
-      let go b = do
-            open <- H.hIsOpen h
-            when open $ do
-              writable <- H.hIsWritable h
-              when writable $ Char8.hPutStrLn h b
-      return go
-    output h trigger = do
+      void $ liftIO $ forkIO $ fix $ \loop -> do
+        newMessage <- atomically $ readTChan channel
+        open <- H.hIsOpen h
+        when open $ do
+          writable <- H.hIsWritable h
+          when writable $ do
+            Char8.hPutStrLn h newMessage
+            loop
+      return (liftIO . atomically . writeTChan channel)
+    output h processFinished trigger = do
       H.hSetBuffering h H.LineBuffering
       let go = do
             open <- H.hIsOpen h
             when open $ do
               readable <- H.hIsReadable h
               when readable $ do
-                out <- BS.hGetSome h 32768
+                out <- BS.hGetNonBlocking h 32768
```

Comment:
We really shouldn't make this a busy wait. Is this problem not solved by checking for `isEOF` here and letting the threads die naturally rather than killing/cancelling them?

Reply:
By "this problem" I mean the entire issue, not just this bit of code.

Reply:
Blocking `hGetSome` *should* get interrupted if the handle dies right?

Reply:
I think so. From the docs for `hGet`:
> If the handle is a pipe or socket, and the writing end is closed, hGet will behave as if EOF was reached.

Reply:
@meditans was there something that caused you to switch to NonBlocking in the first place?

Reply:
You shouldn't even need to check for isEOF

Reply:
So, I thought about that and I welcome alternative solutions, or the reason for which that's not a problem; the issue I was trying to solve making that non-blocking is this:

* the underlying process dies
* it sets the `IORef` that signals that it died
* now the threads have the opportunity of finishing the work on the handle, but we still need a way of saying `we're done with the handles now`, so that we know it's safe to clean the process

 The reason for which I thought is not enough to just use `hGetSome` is that that function blocks on the empty handle. The case I was worried about is this, from the documentation of `hGetSome`:
 `hGetSome only blocks if there is no data available, and EOF has not yet been reached.`

I didn't want to be in a situation in which there's no data available, and no EOF is reached.
I see now that maybe my mental model of how a handle works was wrong; I though that when the underlying process died, there was no way of knowing it looking at the handle. If instead, when the underlying process dies, the handle hand all the data and then simulate an `EOF`, and then terminates, then I suggest to return back to the original code. But anyway, we need to know when this threads terminate so that we complete the cleanup.

@3noch @tomsmalley

Reply:
I think that the problem that this issue solves is that before, if the underlying process terminates, the handles are immediately closed, and there is a lot of data in the handles, we could have lose them; now we tell the threads that handle them that the code terminated, so they have a chance to use all the data in the handles. I don't see how just checking for `isEOF` there would alleviate this concern.

Reply:
> If instead, when the underlying process dies, the handle hand all the data and then simulate an `EOF`, and then terminates, then I suggest to return back to the original code.

I think that is exactly what happens.

~~~~
Prelude> import System.Process
Prelude System.Process> import Data.ByteString
Prelude System.Process Data.ByteString> import System.IO
Prelude System.Process Data.ByteString System.IO> (_,mOut,_,p) <- createProcess $ (shell "echo test") { std_out = CreatePipe }
Prelude System.Process Data.ByteString System.IO> getProcessExitCode p
Prelude System.Process Data.ByteString System.IO> mOut
Just {handle: fd:12}
Prelude System.Process Data.ByteString System.IO> Just h = mOut
Prelude System.Process Data.ByteString System.IO> h
{handle: fd:12}
Prelude System.Process Data.ByteString System.IO> hGetSome h 2
"te"
Prelude System.Process Data.ByteString System.IO> hGetSome h 2
"st"
Prelude System.Process Data.ByteString System.IO> hGetSome h 2
"\n"
Prelude System.Process Data.ByteString System.IO> hGetSome h 2
""
Prelude System.Process Data.ByteString System.IO> hGetSome h 2
""
Prelude System.Process Data.ByteString System.IO> hClose h
Prelude System.Process Data.ByteString System.IO> hGetSome h 2
*** Exception: fd:12: hGetBufSome: illegal operation (handle is closed)
Just ExitSuccess
~~~~

Reply:
@lspitzner Thank you that is exactly the test I wanted to run myself. The only additional check is that `hIsOpen` is still true in all those cases but I can't imagine it's set to false since this is the reading end of the pipe.

Reply:
yeah `hIsOpen h` keeps returning `True` until the `hClose h` invocation.

Reply:
that is, even if EOF is reached, the handle is still open (the reading end only I suppose, but that's what the handle corresponds to).

Reply:
~~~~
Prelude System.Process Data.ByteString System.IO> hIsOpen h
True
Prelude System.Process Data.ByteString System.IO> hGetSome h 2
"\n"
Prelude System.Process Data.ByteString System.IO> hIsOpen h
True
Prelude System.Process Data.ByteString System.IO> hGetSome h 2
""
Prelude System.Process Data.ByteString System.IO> hIsOpen h
True
Prelude System.Process Data.ByteString System.IO> hClose h
Prelude System.Process Data.ByteString System.IO> hIsOpen h
False
~~~~

Reply:
I think I've fixed this.